File: lclang/src/main/java/lclang/libs/reflection/ReflectionValue.java
package lclang.libs.reflection;

import lclang.Link;
import lclang.libs.lang.classes.LibraryClass;
import lclang.libs.lang.classes.StringClass;
import lclang.types.Type;
import lclang.types.Types;

public class ReflectionValue extends LibraryClass {
    public static final String NAME = "ReflectionValue";
    public static final ReflectionValue INSTANCE = new ReflectionValue();
    public static final Type TYPE = INSTANCE.classType;

    public Link value;

    private ReflectionValue() {
        super(NAME);
        constructor = method((caller, lcClasses) -> new ReflectionValue(lcClasses.get(0)
                .executor.getLink(caller, lcClasses.get(1).toString(caller))), Types.ANY, StringClass.type, TYPE);
    }

    public ReflectionValue(Link value) {
        this();

        this.value = value;
        globals.put("type", method((caller, lcClasses) -> StringClass.get(value.type.toString()), StringClass.type));
        globals.put("get", method((caller, lcClasses) -> value.get.invoke(caller), Types.ANY));
        globals.put("set", voidMethod((caller, lcClasses) ->
                        value.set.invoke(caller, lcClasses.get(0)), Types.ANY));
    }
}

File: lclang/src/main/kotlin/lclang/libs/reflection/ReflectionValue.kt
package lclang.libs.reflection

import lclang.Caller
import lclang.LCClass
import lclang.Value
import lclang.libs.lang.NullClass
import lclang.method
import lclang.methods.Method
import lclang.types.Types

class ReflectionValue(value: Value): LCClass("ReflectionValue") {
    init {
        globals["get"] = method(returnType = value.type) { value.get(this) }
        globals["set"] = object : Method(this@ReflectionValue, arrayOf(Types.ANY), Types.VOID) {
            override fun call(caller: Caller, args: List<Value>): Value {
                value.set(caller, args[0])
                return NullClass.NULL.asValue()
            }
        }.asValue()
    }
}